File: python/decidekit/providers/fixture.py
from __future__ import annotations

from copy import deepcopy

from ..errors import ProviderError
from ..types import DecisionQuestion, DecisionProvider, JsonValue, ProviderResponse, Usage


class FixtureProvider:
    name = "fixture"

    def __init__(self, answers: dict[str, dict], model: str = "fixture-v1", latency_ms: int = 0):
        self.answers = answers
        self.model = model
        self.latency_ms = latency_ms

    def evaluate(
        self,
        state: JsonValue,
        questions: dict[str, DecisionQuestion],
        model: str | None = None,
    ) -> ProviderResponse:
        answers: dict[str, dict] = {}
        for decision_id in questions:
            if decision_id not in self.answers:
                raise ProviderError(f'Fixture does not contain an answer for "{decision_id}"')
            answers[decision_id] = deepcopy(self.answers[decision_id])
        return ProviderResponse(
            answers=answers,
            provider=self.name,
            model=model or self.model,
            usage=Usage(cost_usd=0.0),
            latency_ms=self.latency_ms,
        )

File: src/providers/fixture.ts
import { ProviderError } from "../errors.js";
import type { DecisionAnswer, DecisionProvider, DecisionRequest, ProviderResponse } from "../types.js";

export interface FixtureProviderOptions {
  answers: Record<string, DecisionAnswer>;
  name?: string;
  model?: string;
  latencyMs?: number;
}

export class FixtureProvider implements DecisionProvider {
  readonly name: string;
  readonly #answers: Record<string, DecisionAnswer>;
  readonly #model: string;
  readonly #latencyMs: number;

  constructor(options: FixtureProviderOptions) {
    this.name = options.name ?? "fixture";
    this.#answers = options.answers;
    this.#model = options.model ?? "fixture-v1";
    this.#latencyMs = options.latencyMs ?? 0;
  }

  async evaluate(request: DecisionRequest): Promise<ProviderResponse> {
    const answers: Record<string, DecisionAnswer> = {};
    for (const id of Object.keys(request.questions)) {
      const answer = this.#answers[id];
      if (!answer) throw new ProviderError(`Fixture does not contain an answer for "${id}".`);
      answers[id] = structuredClone(answer);
    }
    return {
      answers,
      provider: this.name,
      model: request.model ?? this.#model,
      latencyMs: this.#latencyMs,
      usage: { inputTokens: 0, outputTokens: 0, costUsd: 0 },
    };
  }
}
